로그인 및 페이지 진입 시 Redux user 정보 저장 기능 구현

jma1020 opened the pull request:
# Pull Request
 Feature
## Summary

로그인 후 및 페이지 진입 시 Redux user 정보 저장 추가

## Describe your changes

- user 체크 및 정보를 가져오는 userConfirm api 함수 구현
- 로그인 이후 userConfirm 실행되도록 수정
- 각 페이지 userConfrim 이후 saveUser dispatch 실행

## Issue number and link
#44 

## ETC
/users/profile 요청 시 토큰이 없을 경우 /login 으로 redirect 되는데 백엔드 쪽에서 수정이 필요할 것 같습니다.
백엔드 쪽 수정내용에 따라서 userConfirm return 과 error 형식 변경필요

jma1020 committed:
feature: userSlice data 형태 생성 및 saveUser 수정

- initialState 수정
- saveUser 수정

jma1020 committed:
feature: 로그인 및 페이지 진입 시 user정보 저장 및 로그인 체크

- 유저정보를 가져오는  userConfirm api 함수 구현
- Login api 이후 userConfirm 연결
- _app user정보 없을 경우 userConfirm 함수 실행하도록 작성
- Login, _app-> saveUser dispatch 실행

jma1020 committed:
fix: 불필요한 console 삭제

jma1020 committed:
fix: 불필요한 console 삭제 및 간단한 코드 수정

jma1020 commented on pages/_app.tsx:
```diff
@@ -2,9 +2,25 @@ import type { AppProps } from 'next/app';
+import { useDispatch, useSelector } from 'react-redux';
 
 function MyApp({ Component, pageProps }: AppProps) {
+  const user = useSelector((state: RootState) => state.user);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (!user.isLogIn) {
+      (async function logincheck() {
+        const { data, isLogIn } = await userConfirm();
+        dispatch(saveUser({ isLogIn, ...data }));
+      })();
+    }
+  }, []);
+
```
Login 되었을 경우 계속 실행되는게 비효율적인 것 같아 추가했는데, 다른 괜찮은 형식있으면 피드백 부탁드립니다!

sangrimlee replied:
지금도 저는 충분히 좋아보이는데 회의에도 나왔듯 계속해서 로그아웃된 상태일때, 계속 요청을 하는지 확인해보고 나중에 변경하면 좋을 것 같아요.

jma1020 replied:
넵!! 좋습니다 ㅎㅎ

seongjin2427 commented on api/auth.ts:
```diff
@@ -8,10 +8,39 @@ interface LoginType {
+
+export const userConfirm = async () => {
+  try {
+    const data = await instance.get('/users/profile');
+
+    return {
+      data: data.data,
+      isLogIn: true,
+    };
```
여기서도 살짝 디스트럭쳐링해서
{ data } 로 바로 뽑으면 좋을 것 같습니당 ㅎㅎㅎ

jma1020 replied:
앗 수정하겠습니다!

jma1020 committed:
fix: 간단한 코드 수정

sangrimlee commented on store/slice/userSlice.ts:
```diff
@@ -1,13 +1,24 @@
   reducers: {
     saveUser(state, action) {
-      state.user = action.payload.user;
+      state.authentication = action.payload.authentication == 1 ? true : false;
+      state.name = action.payload.name;
+      state.nickName = action.payload.nickName;
+      state.picture = action.payload.picture;
+      state.isLogIn = action.payload.isLogIn;
```
이 부분을 깔끔하게 state를 object로 재선언할 수 있지 않을까요??

```
state = {
  ...state,
  ...action.payload
}
```
위와 같은 방식으로 할 수 있을 것 같아요.
물론 지금 type 지정이 없어서 오류가 날 것 같긴한데, type지정을 한다면 할 수 있지 않을까요??


그리고 state를 이런 식으로해도 괜찮지 않을까요??

```
interface User {
 authentication: boolean;
 ...
}
 
interface UserState {
 user: User | null;
 isLogIn: boolean;
}
```

왜냐하면 로그아웃할때, user 정보를 initialState로 초기화하는 것보다는 null로 처리하는게 좋아보입니다!!

jma1020 committed:
feat: store 얼리이어스 추가

jma1020 committed:
refact: userSlice state 형식 및 saveUser dispatch 리팩토링 및 type 지정

- userSlice state 형식 변경
- saveUser 리팩토링
- AppDispatch 추가 및 일관성있는 타입으로 공통화

jma1020 committed:
refact: login 및 api 함수 리팩토링

- api 함수 리팩토링 및 변수명 수정
- 로그인페이지 및 _app 리팩토링

jma1020 committed:
feat: 로그아웃 기능 추가

- 로그아웃 userSlice reducer 추가
- logout api 함수 추가
- GNB 컴포넌트 로그아웃 버튼에 logout 기능 구현

jma1020 committed:
refact: axios error 처리 수정

sangrimlee reviewed (approved):
LGTM 👍🏼

jma1020 committed:
fix: 사용하지 않는 코드 삭제

seongjin2427 reviewed (approved):
고생하셨습니다! 😀

jma1020 committed:
fix: 간단한 타입 수정

jma1020 committed:
Merge branch 'develop' into feature/login-redux
